Editor mirror tiles layer

drogoganor opened the pull request:
Added a mirror tile brush to the map editor for marking out tiles. This should allow creators of mirror maps to have an easier time of things.

There's two modes:

**Flip Mode**
Flip mirror tiles along one or both central axes of the map.

![ra-2024-01-27T040425597Z](https://github.com/OpenRA/OpenRA/assets/1639681/c08159c9-f289-497d-b2ad-dfcb23f3def5)

You can supply an axis angle to flip along an angled axis. This angle is in increments of 15 degrees.

**Rotate Mode**
Create mirror tiles by rotating instead of flipping. Can be used for 2-8 player maps.

![ra-2024-01-27T040658120Z](https://github.com/OpenRA/OpenRA/assets/1639681/9d21a54f-0293-4811-b6cb-43c6b184c775)

Mirror tiles are not persisted to the map file.
The layer is enabled by default in the editor, and is toggled with F3.

Potentially needs a clear button and an editable center offset.

Source artwork can be found [here](https://drogoganor.net/mirrortiles.zip).

drogoganor committed:
Editor marker tiles layer

Porenutak commented:
thanks awesome PR again. Here's few notes from user perspective:
- there should be option to delete mirror tiles or reset selected mirror color.
- Axis angle should have visual representation. Some kind of neutral line. So user know against what he is mirroring.

Mailaender commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,255 @@
+		public readonly string Palette = "mirrortiles";
+
+		[Desc("Image to use.")]
+		public readonly string TileImage = "mirrortiles";
```
Why don't you define the mirror counter-parts in the tile set instead?

drogoganor commented:
This needs to go back into the oven. As I was implementing the axis angle display, I realized the isometric mod tile mirroring only has the broad appearance of working, but is actually completely broken. I need to re-do the mirror tile position calculation in a way that respects the isometric projection.

Additionally, I think I need to put the overlay checkboxes i.e. "Show Terrain Geometry", "Show Buildable Terrain" back into a drop-down at the top of the map like it used to be. Then, a drop-down should be added to the Tools panel that lists other tools such as Mirror Tiles. Upon selecting a tool, the controls for that tool should be displayed in the panel. Otherwise, we're completely out of room already.

drogoganor commented:
Current UI design:
![ra-2024-02-08T094519235Z](https://github.com/OpenRA/OpenRA/assets/1639681/3ba33868-e65a-4579-b46b-afcafab5fc4a)

drogoganor commented:
Still requires some chrome fiddling.
Also all the chrome controls are prefixed with MIRROR_ and variable names are prefixed with `mirror` which is a bit verbose.
Probably the mirror layer should be enabled when the user starts painting if it's turned off (it's on by default)
The buttons could probably be disabled if there's no tiles to clear.

PunkPun reviewed (commented):
crash

```
Exception of type `System.IndexOutOfRangeException`: Index was outside the bounds of the array.
   at OpenRA.Mods.Common.Widgets.PaintMirrorTileEditorAction.Add(CPos target) in OpenRA/OpenRA.Mods.Common/EditorBrushes/EditorMirrorLayerBrush.cs:line 138
   at OpenRA.Mods.Common.Widgets.EditorMirrorLayerBrush.HandleMouseInput(MouseInput mi) in /Users/gustas/Documents/GitHub/OpenRA/OpenRA.Mods.Common/EditorBrushes/EditorMirrorLayerBrush.cs:line 76
   at OpenRA.Mods.Common.Widgets.EditorViewportControllerWidget.HandleMouseInput(MouseInput mi) in /OpenRA/OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs:line 89
```

I feel like there should be an erasor brush added, perhaps add a non-colored square, or with an icon

I also recon that people would like to save and import overlays, atm mappers already do this so there is clearly a need

as for the UI, I feel like the axis guide checkbox should just be to the side of the slider instead of in a separate row. Then the overlays dropdown could be moved there as well

<img width="528" alt="Screenshot 2024-02-08 at 14 23 16" src="https://github.com/OpenRA/OpenRA/assets/37534529/7bd9414d-d94f-4402-baee-581239040a30">

the guide lines could be easily scaled to go to the edges of the map

I'll do a code review later

PunkPun commented:
oh, and also for functional purposes I think the overlay should be more transparent. What matters is what is behind the overlay than the overlay itself. I wonder wether we should give control over the transparency to the player 🤔

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Editor/MapToolsLogic.cs:
```diff
@@ -0,0 +1,82 @@
+#region Copyright & License Information
```
it would definitely be better to generate the sprites ingame rather than use art assets. Something like `FillRect` from `RgbaColorRenderer` should do the trick

drogoganor replied:
Will this work for isometric tiles?

PunkPun replied:
you specify the corners of the quad, so yes

PunkPun reviewed (commented):
there seems to be an error with axis calculation in TS

Map: They All Float
<img width="2050" alt="Screenshot 2024-02-27 at 14 35 17" src="https://github.com/OpenRA/OpenRA/assets/37534529/821903a1-9293-4474-83dd-0cb718b5ee5d">

PunkPun commented on OpenRA.Game/Graphics/MirrorTileRenderable.cs:
```diff
@@ -0,0 +1,57 @@
+		public void Render(WorldRenderer wr)
+		{
+			var map = wr.World.Map;
+			var uv = pos.ToMPos(map);
+			var r = map.Grid.Ramps[map.Ramp[uv]];
+			var wpos = map.CenterOfCell(uv.ToCPos(map)) - new WVec(0, 0, r.CenterHeightOffset);
```
```suggestion
			var r = map.Grid.Ramps[map.Ramp[pos]];
			var wpos = map.CenterOfCell(pos) - new WVec(0, 0, r.CenterHeightOffset);
```

PunkPun replied:
the idea was to remove uv

drogoganor replied:
Oh um, sorry about that. I'm still feeling quite ill and didn't quite get that

PunkPun replied:
Suggestions are there to be copypasted 😋

PunkPun commented on OpenRA.Game/Graphics/MirrorTileRenderable.cs:
```diff
@@ -0,0 +1,57 @@
+			var r = map.Grid.Ramps[map.Ramp[uv]];
+			var wpos = map.CenterOfCell(uv.ToCPos(map)) - new WVec(0, 0, r.CenterHeightOffset);
+
+			var corners = r.Corners.Select(corner => wr.Viewport.WorldToViewPx(wr.Screen3DPosition(wpos + corner))).ToArray();
```
```suggestion
			var corners = r.Corners.Select(corner => wr.Viewport.WorldToViewPx(wr.Screen3DPosition(wpos + corner))).ToList();
```

This isn't the cleanest way about it but it does work. A small perf improvement would be casting to list instead. (List takes up more space in memory but the function is faster, since this is a temporary variable we don't need to save memory)

PunkPun reviewed (commented):
I feel like no mirror mode should be added for completeness.  People may want to do corrections

drogoganor commented:
Added a not very thrilling Erase icon:
![image](https://github.com/OpenRA/OpenRA/assets/1639681/22adf8fc-1cb1-4c95-9a14-e16beb1bcad9)
Annoying gap between last button and scroll container.
User can now set Mirror Mode to None so they can mark out arbitrary tiles anywhere

Still to fix:
* Save and load overlays
* Possibly add axis guide checkbox to the same line as Axis Angle slider

PunkPun commented:
related #10513

drogoganor commented:
I've added the ability to persist the mirror tiles layer and settings. It automatically saves to `[SupportDir]\Editor\{mod}\{version}\MirrorTiles\{mapfilename}.json` when you save a map that has any painted mirror tiles. I'm not sure if we like using json but it was definitely convenient.
Not sure if this is what you had in mind but I'm pretty sure it will be suitable for most mapper's needs.

PunkPun commented on OpenRA.Mods.Common/EditorBrushes/EditorMirrorLayerBrush.cs:
```diff
@@ -0,0 +1,161 @@
+			foreach (var paintTile in paintTiles)
+			{
+				mirrorLayerOverlay.SetTile(paintTile.Cell, type);
+			}
```
style nit: remove braces

PunkPun commented on OpenRA.Mods.Common/EditorBrushes/EditorMirrorLayerBrush.cs:
```diff
@@ -0,0 +1,161 @@
+			foreach (var paintTile in paintTiles)
+			{
+				mirrorLayerOverlay.SetTile(paintTile.Cell, paintTile.Previous);
+			}
```
style nit: remove braces

PunkPun commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,442 @@
+	[TraitLocation(SystemActors.EditorWorld)]
+	public class MirrorLayerOverlayInfo : TraitInfo
+	{
+		[Desc("Swatch colors to display in the mirror tile selection box.")]
```
```suggestion
		[Desc("A list of colors to be used for drawing.")]
```

perhaps?

PunkPun commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,442 @@
+			Color.FromArgb(255, 0, 220),
+		};
+
+		[Desc("Alpha blend amount.")]
```
```suggestion
		[Desc("Default alpha blend.")]
```

PunkPun commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,442 @@
+			for (var i = 0; i < Info.Colors.Length; i++)
+			{
+				alphaBlendColors[i] = Color.FromArgb(tileAlpha, Info.Colors[i]);
+			}
```
style nit: remove braces

PunkPun commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,442 @@
+		{
+			if (Tiles.TryGetValue(tileType, out var set))
+			{
+				foreach (var pos in new HashSet<CPos>(set))
```
```suggestion
				foreach (var pos in set)
```

also remove braces

PunkPun commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,442 @@
+
+		public void ClearAll()
+		{
+			foreach (var tile in Tiles.Keys)
+			{
+				foreach (var position in Tiles[tile])
+				{
+					CellLayer[position] = null;
+				}
+			}
```
```suggestion
			foreach (var position in Tiles.SelectMany(x => x.Value))
				CellLayer[position] = null;
```

PunkPun commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,442 @@
+		{
+			ClearAll();
+
+			foreach (var tile in newTiles.Keys)
+			{
+				var set = new HashSet<CPos>();
+				Tiles.Add(tile, set);
+
+				foreach (var position in newTiles[tile])
+				{
+					set.Add(position);
+					CellLayer[position] = tile;
+				}
+			}
```
```suggestion
			foreach (var type in newTiles)
			{
				var set = new HashSet<CPos>();
				Tiles.Add(type.Key, set);

				foreach (var position in type.Value)
				{
					set.Add(position);
					CellLayer[position] = type.Key;
				}
			}
```

PunkPun commented on OpenRA.Mods.Common/Traits/World/MirrorLayerOverlay.cs:
```diff
@@ -0,0 +1,442 @@
+
+		public void SetSelected(int tile, HashSet<CPos> newTiles)
+		{
+			foreach (var pos in Tiles[tile])
+			{
+				SetTile(pos, null);
+			}
+
+			Tiles[tile].Clear();
+
+			foreach (var pos in newTiles)
+			{
+				Tiles[tile].Add(pos);
+				CellLayer[pos] = tile;
+			}
```
```suggestion
			var type = Tiles[tile];
			foreach (var pos in type)
				SetTile(pos, null);

			type.Clear();

			foreach (var pos in newTiles)
			{
				type.Add(pos);
				CellLayer[pos] = tile;
			}
```

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Editor/MapMirrorTilesLogic.cs:
```diff
@@ -0,0 +1,341 @@
+			dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 150, options, SetupItem);
+		}
+
+		class ClearSelectedMirrorTilesEditorAction : IEditorAction
```
I don't think that editor actions should be in widget logic. Please move them to the brush

PunkPun commented on mods/cnc/chrome/editor.yaml:
```diff
@@ -283,8 +284,8 @@ Container@EDITOR_WORLD_ROOT:
-							Width: 61
-							Height: 50
+							Width: 60
+							Height: 49
```
rebase errors?

PunkPun reviewed (commented):
Overall the PR is in good condition, just some polish is missing. I also want to mention some scope creep

The overlay saving definitely needs to have separate save functionality. The most common usecase is having pre-made overlays that you just import when working on a new map. It means that different map sizes should be handled, centring the overlay when it gets cutoff or is incomplete

It will still be much more polished if the symmetry line was clamped to map edges
<img width="1051" alt="Screenshot 2024-03-09 at 22 12 00" src="https://github.com/OpenRA/OpenRA/assets/37534529/77e7be86-010d-43e4-8845-93cd62075617">

Porenutak commented:
The symetry line looks broke:


https://github.com/OpenRA/OpenRA/assets/100044015/b13a41e8-3ff5-484e-998c-7adfae20f4d7


I also agree that mirror should start from map edges. Some maps may have lots of unreachable area that is asymmetric:
![d2k-2024-03-14T092157513Z](https://github.com/OpenRA/OpenRA/assets/100044015/d3a1a6e2-f578-437f-91ae-5f3149da9a81)

drogoganor commented:
@Porenutak , you'll need to do a `git pull` and `make all` on this branch to get the latest code.

Porenutak commented:
pls someone help me:

![hmm](https://github.com/OpenRA/OpenRA/assets/100044015/231ef7d1-d52d-498b-ac9b-bf4158ade404)

Im little confused. Why github wanna push drogoganor  commits from my branch copy? I didnt change anything it the code.

Porenutak commented:
Dont know how hard is this to implement, but to speed up workflow it would be nice to switch to eraser via modifier. Like with LMB I wil add mirror tiles, but with CTRL+ LMB I will erase them.

PunkPun commented:
I dunno if that would be better. IMO we should keep the current workflow for this PR

Porenutak commented:
Its more a thought for the future. Using modifiers is always faster that clicking on tab icons. All modern editors works this way.

drogoganor commented:
We might be fresh out of room on the panel. I discovered when I tried to add controls to save and load the mirror tiles layer, that the scrollbar doesn't enable. I think `ScrollPanelWidget` doesn't like being pre-populated with child controls from the yaml. Setting the ContentHeight doesn't enable the scrollbar, either. I think it has to be manually populated with `ScrollItemWidget`s at runtime for scrolling to be enabled.

PunkPun commented:
See how the settings panels work

drogoganor commented:
OK yeah, I got that figured out.

![ra-2024-03-18T143559390Z](https://github.com/OpenRA/OpenRA/assets/1639681/8fd798f9-8629-4cbe-8c35-b7eb82412e26)

There's a fair bit of work to do on this part so it's foolproof. Needs a validation label like on the Actor ID edit.

A potential concern is that the user could easily overwrite another mirror tile file, since there's no confirmation dialog.
Also argh-worthy is yielding focus from the text input.

drogoganor commented:
After some consideration, I think it might be better to omit the custom mirror tile file save/load feature, and just automatically persist to a file that matches the map filename instead.
I think the use-case for loading mirror tiles created on other maps is maybe limited. Also I'm worried about users accidentally overwriting a mirror tile file for another map by accident. In general I'm reluctant to have any file management code in the editor other than the map file itself.
Pre-made grids like this:
![image](https://github.com/OpenRA/OpenRA/assets/1639681/50c553f4-b38e-4fe6-8e7b-511c8aa19180)
Might have limited utility now since the user can simply paint tiles and have them mirrored as they require.

PunkPun commented:
We don't necessarily need ingame file management. But the game should handle imports from maps of other sizes. Currently I get this crash

```
Exception of type `System.IndexOutOfRangeException`: Index was outside the bounds of the array.
   at OpenRA.CellLayer`1.get_Item(CPos cell) in OpenRA/OpenRA.Game/Map/CellLayer.cs:line 78
   at OpenRA.Graphics.MirrorTileRenderable.Render(WorldRenderer wr) in OpenRA/OpenRA.Game/Graphics/MirrorTileRenderable.cs:line 45
   at OpenRA.Graphics.WorldRenderer.DrawAnnotations() in OpenRA/OpenRA.Game/Graphics/WorldRenderer.cs:line 337
   ```
   
 I suppose fixing the TS axis line would be too much of a scope screep. But I wanted to still mention that it's lieing
 
 <img width="1009" alt="Screenshot 2024-03-20 at 16 04 53" src="https://github.com/OpenRA/OpenRA/assets/37534529/45c8ebbc-6d23-4cc0-ad3f-68081e9b26e7">

PunkPun commented on mods/cnc/chrome/editor.yaml:
```diff
@@ -366,8 +367,8 @@ Container@EDITOR_WORLD_ROOT:
-							Width: 61
-							Height: 75
+							Width: 60
+							Height: 73
```
rebase error?

drogoganor replied:
Regarding that, I found the size of a panel in the editor is slightly incorrect, and thought I would fix it while I was here. But I think I might back it out instead. Thanks

Porenutak commented:
> We don't necessarily need ingame file management. But the game should handle imports from maps of other sizes. 

Can u give at least one example why anyone needs to import mirror tiles from other map?
I dont see the point.

drogoganor commented:
I think making the mirror tile file work if you manually copy and rename it from another map would be a good idea. I'll get onto that.

Yes, that axis line issue in TS is such a pain. I thought it was due to elevation but apparently not.

drogoganor commented:
* I renamed the feature to Marker Tiles instead of Mirror Tiles
* Now defaults to Mirror Mode: None
* The "Show Axis Guide" checkbox is now removed (it simply always shows in Flip mode, unless the Marker Tiles overlay is disabled)
* Renaming a marker tiles file to the name of another map of a different size will no longer cause a crash (tiles that aren't on the map will be dropped)
* Fixed cnc editor chrome regression

PunkPun commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0f72a0c23c52aa4efa1be117886b6e6bbd8b9e2d...f57605b00c1a2fe39ed3a24de4a25d121859739e)
